server/src/main/java/org/candlepin/audit/EventSinkImpl.java

```diff
@@ -127,8 +127,8 @@ public void queueEvent(Event event) {
 
         try {
             // Lazily initialize the message sender when the first
-            // message gets queued.
-            if (messageSender == null) {
+            // message gets queued or initialize session if it's closed.
+            if (messageSender == null || messageSender.session.isClosed()) {
```

Comment:
This works, but is still breaking the encapsulation open to accomplish the goal. The purpose of the ```EventMessageSender``` is to be our abstraction layer for managing the session, so we should avoid spinning up a new one to do the work it should be doing. From the other change in this code, it appears you've already spotted the downside in doing it this way: you need to do the close check in every operation; making the session management exist in three places, rather than one.

However, this does highlight another minor problem with the existing code: EventMessageSender isn't static. The wrapper does not care about the state of its outer class, nor should it ever use it. We don't need the baggage an inner class inherits in such a case, and we should be declaring it statically.

Reply:
Right, made some adjustments to it. 
Checks on `sendMessages` & `cancelMessages` methods are to prevent doing anything if we don't have session opened. With [this line](https://github.com/candlepin/candlepin/blob/master/server/src/main/java/org/candlepin/async/JobManager.java#L1286) inside `JobManager` which make sure that events are always sent out or rolled backed (in case of exception). Now with transnational wrapper specifically in HypervisorUpdateJob the AMQ session get closed, which leads to problem when control goes back to JobManager.